### PR TITLE
factorial challenges

### LINKED ISSUE #58: factorial
I implement  of the factorial function which designed to compute the factorial of non -negative integer recursively .

### PR DESCRIPTION
**Pull request by Safaishag**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/factorial.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        1
+        >>> factorial(5)
+        120
+        >>> factorial(-1)
+        Traceback (most recent call last):
+            ...
+        ValueError: Input must be a non-negative integer.
+    """
+    # Defensive assertion: Input must be a non-negative integer
+    if not isinstance(n, int):
```
Comment (Goutbi) [inline solutions/factorial.py:41]: Consider adding a brief comment to explain why type checking is necessary here for clarity but overall it's clear

File: solutions/tests/test_factorial.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    def test_negative_input(self):
+        """
+        Test that negative inputs raise a ValueError.
+        """
+        with self.assertRaises(ValueError):
+            factorial(-5)
+
+    def test_non_integer_input(self):
+        """
+        Test that non-integer inputs raise a TypeError.
```
Comment (Goutbi) [inline solutions/tests/test_factorial.py:48]: This is a solid test for TypeError. Testing string inputs like factorial('5') could further strengthen coverage.

**Review by Goutbi — APPROVED** [review]
The code and tests are well-written, covering all critical functionality with robust defensive assertions.
